### PR TITLE
feat: add depth flag

### PR DESCRIPTION
**Pull request by tomberek**
limit recursion to a default depth of 1

needs improvement: recursion is only the number of "."

Fixes: #10 #9

### CODE REVIEW

File: src/nix-eval-jobs.cc [hunk 1]
```diff
@@ -231,17 +241,24 @@ static void worker(
...
-                for (auto & i : v->attrs->lexicographicOrder()) {
-                    std::string name(i->name);
-                    if (name.find('.') != std::string::npos || name.find(' ') != std::string::npos) {
-                        printError("skipping job with illegal name '%s'", name);
-                        continue;
+                // Hacky way to get depth.... fails with embedded "."
+                size_t n = std::count(attrPath.begin(), attrPath.end(), '.')+(attrPath == "" ? 0: 1);
+                if (n < myArgs.depth){
+                    auto attrs = nlohmann::json::array();
+                    Bindings::iterator j = v->attrs->find(state.sRecurseForDerivations);
```
Comment (Mic92) [inline src/nix-eval-jobs.cc:248]: Can you make a test case for this or just extend this file? https://github.com/nix-community/nix-eval-jobs/blob/4bab9e449da80b58507d8d3ab6c051701e8d65e4/tests/assets/ci.nix#L6

Comment (Mic92) [reply src/nix-eval-jobs.cc:248]: Otherwise looks good to me.

File: src/nix-eval-jobs.cc [hunk 2]
```diff
@@ -59,6 +60,15 @@ struct MyArgs : MixEvalArgs, MixCommonArgs
             }},
         });
 
+        addFlag({
+            .longName = "depth",
```
Comment (Mic92) [inline src/nix-eval-jobs.cc:64]: I guess this is nice to preserve backwards compatibility but it should be also documented in the README.

**Comment by Mic92** [comment]
In general. Have you considered re-using: https://github.com/NixOS/nix/blob/d5d4d980427aca3849b90bfe1694b6d1d14532fb/src/libexpr/get-drvs.cc#L337
and if yes, what where the reasons not to copy the code?
The reason its not here used where because this was copied from hydra.

**Comment by tomberek** [comment]
I copied the logic and code from “getDerivations”, but not using the function itself because I believed the intent is to keep using the parallel server/worker model on a per-attribute basis. “getDerivations” does a single threaded depth-first-search while this (and hydra) does a breadth-first parallel search.

**Comment by adisbladis** [comment]
I really don't like to have a `depth` flag that defaults to some arbitrary low value since that wouldn't work well for evaluating trees.
I've gone ahead and implemented support for _only_ `recurseForDerivations` in #62. 

A depth flag should be completely separate and optional.

**Review by adisbladis — CHANGES_REQUESTED** [review]
This flag needs to be completely optional and not default to some arbitrary low value that would break tree evaluation.

File: src/nix-eval-jobs.cc [hunk 3]
```diff
@@ -231,17 +241,24 @@ static void worker(
...
               {
-                auto attrs = nlohmann::json::array();
-                StringSet ss;
-                for (auto & i : v->attrs->lexicographicOrder()) {
-                    std::string name(i->name);
-                    if (name.find('.') != std::string::npos || name.find(' ') != std::string::npos) {
-                        printError("skipping job with illegal name '%s'", name);
-                        continue;
+                // Hacky way to get depth.... fails with embedded "."
+                size_t n = std::count(attrPath.begin(), attrPath.end(), '.')+(attrPath == "" ? 0: 1);
```
Comment (adisbladis) [inline src/nix-eval-jobs.cc:245]: This is fine for now and would be addressed by #43.

Comment (adisbladis) [inline src/nix-eval-jobs.cc:245]: This is fine for now and would be addressed by #43.

**Review by adisbladis — CHANGES_REQUESTED** [review]

**Comment by Mic92** [comment]
This PR has been stalled. Please re-open if you plan to continue working on this.

**Pull request closed without merging**